flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/AbstractUserClassPathJobGraphRetriever.java

```diff
@@ -0,0 +1,117 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.entrypoint.component;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.FileVisitOption;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ *  Abstract class for the JobGraphRetriever, which wants to get classpath user's code depends on.
+ */
+public abstract class AbstractUserClassPathJobGraphRetriever implements JobGraphRetriever {
+
+	protected static final Logger LOG = LoggerFactory.getLogger(AbstractUserClassPathJobGraphRetriever.class);
+
+	/** The directory contains all the jars, which user code depends on. */
+	@Nullable
+	private final String jobDir;
+
+	private List<URL> userClassPaths;
+
+	public AbstractUserClassPathJobGraphRetriever(String jobDir) {
+		this.jobDir = jobDir;
+	}
+
+	protected List<URL> getUserClassPaths() throws IOException {
+		if (userClassPaths == null) {
+			userClassPaths = getRelativeJarsURLFromDir(jobDir);
+		}
+		return userClassPaths;
+	}
+
+	/**
+	 * Scan all the jar files in the {@code dir} and return all these jar files' relative URLs to "user.dir".
+	 * @param dir the dir needed to scan the jar files
+	 * @return the jar files' relative URLs
+	 * @throws IOException
+	 */
+	private List<URL> getRelativeJarsURLFromDir(String dir) throws IOException {
+
+		if (dir == null) {
+			return Collections.emptyList();
+		}
+
+		if (!Files.exists(Paths.get(dir))) {
+			throw new IllegalArgumentException("the job dir " + dir + " dose not exists.");
+		}
+		if (!Files.isDirectory(Paths.get(dir))) {
+			throw new IllegalArgumentException("the job dir " + dir + " is not a directory.");
+		}
+
+		Path dirPath;
+		final List<URL> jarURLs = new LinkedList<>();
+
+		if (Paths.get(dir).isAbsolute()) {
+			dirPath = Paths.get(System.getProperty("user.dir")).relativize(Paths.get(dir));
```

Comment:
I would suggest to pass in the working directory as a parameter to this method instead of relying on system properties.

Reply:
I think the `System.getProperty("user.dir")` is a standard method to get the working directory in jvm.  In the end, there is always some method to depend on it.  Is there any difference?
Would you give me some hint? thanks.

Reply:
It is ok to use `System.getProperty("user.dir")` for the current working directory. But this should not be the concern of `getRelativeJarsURLFromDir`. Instead this should be decided somewhere else. The method should simply get as an argument the current working directory. That way it is easier to test.

Ideally, we would actually have two methods: One for listing the files, and another to relativize paths wrt to another path. That way we would not have a single function which does multiple things.

Reply:
I got what you mean.